File: QytetetRuby/lib/especulador.rb
#encoding: utf-8
# Francisco Beltrán Sánchez
# Pablo Cordero Romero
module ModeloQytetet
  class Especulador < Jugador
    attr_accessor :fianza

    def self.copia(jugador, fza)
      @fianza
      salida = super(jugador)
      salida.fianza = fza
      
      salida
    end
    
    def pagarImpuesto
      modificarSaldo(-(@casillaActual.precioCompra)/2)
    end

    def deboIrACarcel
      salida = (super && !pagarFianza)

      salida
    end

    def convertirme(fza)
      self
    end

    def pagarFianza
      salida = @saldo>=@fianza

      if salida
        modificarSaldo(-@fianza)
      end
      
      salida
    end
    
    def puedoEdificarCasa(titulo)
      salida = titulo.numCasas<8
      
      salida
    end

    def puedoEdificarHotel(titulo)
      salida = (titulo.numHoteles<8 && titulo.numCasas>=4)

      salida
    end

    def to_s
      "\n-----\nEspeculador: #{super} \n\tFianza= #{@fianza}"
    end
    
    private :pagarFianza
  end
end

File: QytetetJava/src/modeloqytetet/Especulador.java
/**
 * Practica de:
 * 
 * Francisco Beltrán Sánchez
 * Pablo Cordero Romero
 */
package modeloqytetet;

public class Especulador extends Jugador{
    private int fianza;

    protected Especulador(Jugador jugador, int fza){
        super(jugador);
        this.fianza = fza;
    }

    @Override
    protected void pagarImpuesto(){
        super.modificarSaldo(-super.getCasillaActual().getCoste()/2);
    }
    
    @Override
    protected boolean deboIrACarcel(){
        return (super.deboIrACarcel() && !pagarFianza());
    }

    @Override
    protected Especulador convertirme(int fza){
        return this;
    }
    
    private boolean pagarFianza(){
        boolean salida = super.getSaldo() >= fianza;
        
        if(salida){
            super.modificarSaldo(-fianza);
        }
        
        return salida;
    }
    
    @Override
    protected boolean puedoEdificarCasa(TituloPropiedad titulo){
        int numCasas = titulo.getNumCasas();

        return numCasas<8;
    }
    
    @Override
    protected boolean puedoEdificarHotel(TituloPropiedad titulo){
        int numHoteles = titulo.getNumHoteles();
        int numCasas = titulo.getNumCasas();

        return (numHoteles<8 && numCasas>=4);
    }

    @Override
    public String toString() {
        return "\n-----\nEspeculador:" + super.toString() + "\n\tfianza=" + fianza;
    }
}
